pcsx2/SPU2/SndOut_Cubeb.cpp

```diff
@@ -0,0 +1,337 @@
+
+class Cubeb : public SndOutModule
+{
+private:
+	//////////////////////////////////////////////////////////////////////////////////////////
+	// Configuration Vars (unused still)
+	bool m_COMInitializedByUs = false;
+	bool m_SuggestedLatencyMinimal = false;
+	int m_SuggestedLatencyMS = 20;
+	std::string m_Backend;
+
+	//////////////////////////////////////////////////////////////////////////////////////////
+	// Instance vars
+	u64 writtenSoFar = 0;
+	u64 writtenLastTime = 0;
+	u64 positionLastTime = 0;
+
+	u32 channels = 0;
+	cubeb* m_context = nullptr;
+	cubeb_stream* stream = nullptr;
+	bool m_paused = false;
+
+	//////////////////////////////////////////////////////////////////////////////////////////
+	// Stuff necessary for speaker expansion
+	class SampleReader
+	{
+	public:
+		virtual void ReadSamples(void* outputBuffer, long frames) = 0;
+	};
+
+	template <class T>
+	class ConvertedSampleReader final : public SampleReader
+	{
+		u64* written;
+
+	public:
+		ConvertedSampleReader(u64* pWritten)
+			: written(pWritten)
+		{
+		}
+
+		void ReadSamples(void* outputBuffer, long frames) override
+		{
+			T* p1 = (T*)outputBuffer;
+
+			while (frames > 0)
+			{
+				const long frames_to_read = std::min<long>(frames, SndOutPacketSize);
+				SndBuffer::ReadSamples(p1, frames_to_read);
+				p1 += frames_to_read;
+				frames -= frames_to_read;
+			}
+
+			(*written) += frames;
+		}
+	};
+
+	void DestroyContextAndStream()
+	{
+		if (stream)
+		{
+			cubeb_stream_stop(stream);
+			cubeb_stream_destroy(stream);
+			stream = nullptr;
+		}
+
+		if (m_context)
+		{
+			cubeb_destroy(m_context);
+			m_context = nullptr;
+		}
+
+#ifdef _WIN32
+		if (m_COMInitializedByUs)
+		{
+			CoUninitialize();
+			m_COMInitializedByUs = false;
+		}
+#endif
+	}
+
+	SampleReader* ActualReader = nullptr;
+
+public:
+	Cubeb() = default;
+
+	~Cubeb()
+	{
+		DestroyContextAndStream();
+	}
+
+	bool Init() override
+	{
+		ReadSettings();
+
+#ifdef _WIN32
+		HRESULT hr = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+		m_COMInitializedByUs = SUCCEEDED(hr);
+		if (FAILED(hr) && hr != RPC_E_CHANGED_MODE)
+		{
+			ConLog("Failed to initialize COM");
+			return false;
+		}
+#endif
+
+		int rv = cubeb_init(&m_context, "PCSX2", m_Backend.empty() ? nullptr : m_Backend.c_str());
+		if (rv != CUBEB_OK)
+		{
+			ConLog("Could not initialize cubeb context: %d", rv);
+			return false;
+		}
+
+		switch (numSpeakers) // speakers = (numSpeakers + 1) *2; ?
+		{
+			case 0:
+				channels = 2;
+				break; // Stereo
+			case 1:
+				channels = 4;
+				break; // Quadrafonic
+			case 2:
+				channels = 6;
+				break; // Surround 5.1
+			case 3:
+				channels = 8;
+				break; // Surround 7.1
+			default:
+				channels = 2;
+				break;
+		}
+
+		switch (channels)
+		{
+			case 2:
+				ConLog("* SPU2 > Using normal 2 speaker stereo output.\n");
+				ActualReader = new ConvertedSampleReader<StereoOut16>(&writtenSoFar);
```

Comment:
I don't see a `delete` anywhere, any risk this could end up leaking?

Reply:
True, this can be a unique_ptr.